design/kata-api-design.md

```diff
@@ -0,0 +1,112 @@
+# Kata API Design
+To fulfill the [kata design requirements](kata-design-requirements.md), and based on the disscusion on [Virtcontainers API extentions](https://docs.google.com/presentation/d/1dbGrD1h9cpuqAPooiEgtiwWDGCYhVPdatq7owsKHDEQ), the kata runtime library features following APIs:
+-  Sandbox based top API
+-  Storage and network hotplug API
+-  Plugin frameworks for external proprietary Kata runtime extensions
+-  Built-in shim and proxy types and capabilities
+
+## Sandbox Based API
+### Sandbox Management API
+|Name|Description|
+|---|---|
+|CreateSandbox(SandboxConfig)| Create, start a sandbox and return the sandbox structure.|
+|FetchSandbox(ID)| Connect to an existing sandbox and return the sandbox structure.|
+|ListSandbox()| List all existing standboxes with status. |
+
+### Sandbox Operation API
+|Name|Description|
+|---|---|
+|sandbox.Pause()| Pause the sandbox.|
+|sandbox.Resume()| Resume the paused sandbox.|
+|sandbox.Delete()| Destroy the sandbox and remove all persistent metadata.|
+|sandbox.Status()| Get the status of the sandbox and containers.|
+|sandbox.Monitor()| Return a context handler for caller to monitor sandbox callbacks such as error termination.|
+|sandbox.CreateContainer()| Create new container in the sandbox.|
+|sandbox.DeleteContainer()| Delete a container from the sandbox.|
+|sandbox.StartContainer()| Start a container in the sandbox.|
+|sandbox.StatusContainer()| Get the status of a container in the sandbox.|
+|sandbox.EnterContainer()| Run a new process in a container.|
+|sandbox.WaitProcess()| Wait on a process.|
+### Sandbox Hotplug API
+|Name|Description|
+|---|---|
+|sandbox.AddStorage()| Add new storage to the sandbox.|
+|sandbox.AddNetwork()| Add new nic to the sandbox.|
+|sandbox.ListNetwork()| List all nics and their configurations in the sandbox.|
+|sandbox.UpdateNetwork()| Update the configuration of an existing network interface.|
+|sandbox.UpdateRoute()| Update the sandbox route table, e.g. for portmapping support.|
```

Comment:
I think this can be reduced to a more simple API:
- `AddStorage()`
- `AddNetwork()` (covering both `AddRoute` and `AddInterface`)
- `UpdateNetwork()` (covering both `UpdateRoute` and `UpdateInterface`)

Why do you need the list of networks with `ListNetwork()` ? CRI should already tell you what to add or what to update, right ?

Reply:
It is to support CLI functionality like `runv network ls` as requested by @WeiZhang555.

And about combining API for Interface and route table, if you think about the arguments, they will have no overlap between each other. Are you suggesting that we use some kind of a union to combine them?

Reply:
I might need to describe more about CLI interface of `runv/kata network add` especially scenarios

For current runv network support(as I know cc is similar, because I know runv better I'll take runv as example), runv will create a network namespace for runv-shim and docker will treat runv shim as a container. When a CNI plugin inserts a network interface into the "container"'s ns, runv can detect the interface with a nslistener daemon and try to create a tap and copy all configurations from veth to tap, this is how runv's CNI network worked initially.

That leads to some problems when runv is doing this "smart" things for user. First it's hard to detects and copy all the configurations, for example, if the CNI veth is put into an ovs bridge, runv will have to detects which ovs bridge the veth is in and its vlan port and other features(e.g. TSO feature is on or off), that brings lots of hard work to runv and make the "smart" things not that cool any more. What's worse, if user has a custom virtual switch that runv doesn't recognize(we happened to have one), it can do nothing, we have to break runv code and add more customized codes directly into runv.

So let's check how CNI works, a CNI plugin will insert an network interface and add it to some bridge and do more works, neither docker nor runc is aware of this. To imitate this, we contribute a `runv interface` command, if CNI plugin trys to put an network device into "VM"s namespace, it calls into `runv interface add` and the tap is inserted, and CNI plugin can do anything on host and runv doesn't need to care about anything. This is a more direct way than the "listener" and "detect" things.

For `ListNetwork` interface, CNI plugin can double check if the interface is configured correctly, even AddNetwork succeed we don't know if we truly get what we want, or maybe AddNetwork failed partly but we still get enough configuration to make CNI network work.

Reply:
So `ListNetwork` is mostly for a CNI plugin to double check if the resulting network configuration is the same as expected, and thus it can fix any possible middle states, right?

Reply:
@WeiZhang555 Thank you so much for this very detailed explanation. This makes a lot of sense and I actually like the design :)

Reply:
Much like how our discussion for gRPC went, wouldn’t it be better to return status and state if the add network fails?

Reply:
@bergwolf @egernst 

Yes, currently `ListNetwork` is only for double check, currently our CNI plugin will do a retry when `ListNetwork` returns unexpected result. Another usage is when end user(not CNI plugin) run an `AddNetwork` cmd, they generally would like to `ListNetwork` to check how many network interfaces he got in the end without calling into an `docker exec`.

For a CNI plugin, if AddNetwork can return final state of network interfaces, we don't need `ListNetwork`. But from an end user's perspective, a `ListNetwork` will be useful and friendly :-)

Reply:
Sounds good to me. Thanks for the explanation @WeiZhang555

Reply:
@jodh-intel @egernst PTAL and approve the PR if you're fine with it.